packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/MrwSerializationTypeProvider.cs

```diff
@@ -345,6 +374,94 @@ private MethodBodyStatement[] BuildJsonModelWriteCoreMethodBody()
+            {
+                var rawDataType = new CSharpType(typeof(Dictionary<string, BinaryData>));
+                // IDictionary<string, BinaryData> serializedAdditionalRawData = default;
+                additionalRawDataDictionaryDeclaration = Declare(
+                    _privateAdditionalRawDataPropertyType,
+                    AdditionalRawDataVarName,
+                    new DictionarySnippet(_privateAdditionalRawDataPropertyType.Arguments[0], _privateAdditionalRawDataPropertyType.Arguments[1], Default),
+                    out additionalRawDataDictionary);
+                // Dictionary<string, BinaryData> rawDataDictionary = new Dictionary<string, BinaryData>();
+                rawDataDictionaryDeclaration = Declare(
+                       "rawDataDictionary",
+                       new DictionarySnippet(rawDataType.Arguments[0], rawDataType.Arguments[1],
+                       New.Instance(rawDataType)),
+                       out rawDataDictionary);
+                // serializedAdditionalRawData = rawDataDictionary;
+                assignRawData = additionalRawDataDictionary.Assign(rawDataDictionary).Terminate();
+            }
+
+            var propertyDeclarationStatements = GetPropertyVariableDeclarations(out var propertyVarReferences);
+            var assignOptionsIfNull = _mrwOptionsDeserializationParameterSnippet.Assign(ModelSerializationExtensionsSnippet.Wire, true).Terminate();
+            IfStatement valueKindEqualsNullCheck = new(_jsonElementParameterSnippet.ValueKindEqualsNull())
+            {
+                Return(Null)
+            };
+
+            return
+            [
+                assignOptionsIfNull,
+                EmptyLineStatement,
+                valueKindEqualsNullCheck,
+                propertyDeclarationStatements,
+                additionalRawDataDictionaryDeclaration,
+                rawDataDictionaryDeclaration,
+                BuildDeserializePropertiesForEachStatement(rawDataDictionary, propertyVarReferences),
+                assignRawData,
+                Return(New.Instance(_model.Type, GetSerializationCtorParameterValues(additionalRawDataDictionary, propertyVarReferences)))
+            ];
+        }
+
+        /// <summary>
+        /// Gets the variable declarations for the model properties.
+        /// The variable references are outputted in a dictionary of property provider to their corresponding
+        /// variable expression.
+        /// </summary>
+        /// <returns>The property declaration statements.</returns>
+        private MethodBodyStatement[] GetPropertyVariableDeclarations(out Dictionary<PropertyProvider, VariableExpression> propertyVarReferences)
+        {
+            var propertyCount = _model.Properties.Count;
+            MethodBodyStatement[] propertyDeclarationStatements = new MethodBodyStatement[propertyCount];
+            propertyVarReferences = new Dictionary<PropertyProvider, VariableExpression>(propertyCount);
```

Comment:
instead of the extra dictionary here can we add an AsVariableExpression method to PropertyProvider?